pandas/tests/io/parser/mangle_dupes.py

```diff
@@ -40,3 +40,25 @@ def test_thorough_mangle(self):
         df = self.read_csv(StringIO(data), sep=",", mangle_dupe_cols=True)
         assert list(df.columns) == ["a", "a.1", "a.3", "a.1.1",
                                     "a.2", "a.2.1", "a.3.1"]
+
+    def test_thorough_mangle_names(self):
+        # see gh-17095
+        data = "a,b,b\n1,2,3"
+        names = ["a.1", "a.1", "a.1.1"]
+        df = self.read_csv(StringIO(data), sep=",", names=names,
+                           mangle_dupe_cols=True)
+        assert list(df.columns) == ["a.1", "a.1.1", "a.1.1.1"]
```

Comment:
There is also another behaviour that *might* be sensible: a expected result of `"a.1", "a.1.1.1", "a.1.1"]` (so taking into account all names and not only the ones that already were processed). This ends up with only one column being different than specified, while in the test two columns did change. 
Not sure this is necessarily better, just pointing out the possibility. But less changed names sounds like a nice argument. 

I would also add a test with a different order to be specific about this, for example for `names=['a', 'a.1', 'a']`